test/MatchingTests.cpp

```diff
@@ -316,37 +319,36 @@ namespace Simulator
         bool valid = true;
         valid = valid && lte(order.validSince, timestamp);
         valid = valid && lte(timestamp, order.validUntil);
-
         valid = valid && !((order.buy == FieldT::zero()) && (order.allOrNone == FieldT::one()) && lt(fillAmountS, order.amountS));
         valid = valid && !((order.buy == FieldT::one()) && (order.allOrNone == FieldT::one()) && lt(fillAmountB, order.amountB));
         valid = valid && checkFillRate(order.amountS, order.amountB, fillAmountS, fillAmountB);
         valid = valid && (fillAmountS != FieldT::zero());
         valid = valid && (fillAmountB != FieldT::zero());
-
         return valid;
     }
 
-    Settlement settle(const FieldT& timestamp, const OrderState& orderStateA, const OrderState& orderStateB)
+    Settlement settle(const FieldT& timestamp, const OrderState& orderStateA, const OrderState& orderStateB, const FieldT& maxFillS_A, const FieldT& maxFillS_B)
     {
-        Fill fillA = getMaxFillAmounts(orderStateA);
-        Fill fillB = getMaxFillAmounts(orderStateB);
+        Fill fillA = getMaxFillAmounts(orderStateA, maxFillS_A, maxFillS_B);
+        Fill fillB = getMaxFillAmounts(orderStateB, maxFillS_B, maxFillS_A);
 
         bool matchable;
         if (orderStateA.order.buy == FieldT::one())
         {
+            // maker.B < taker.S ==> B.B < A.S
             matchable = match(orderStateA.order, fillA, orderStateB.order, fillB);
             fillA.S = fillB.B;
         }
         else
         {
+            // maker.B < taker.S ==> A.B < B.S
             matchable = match(orderStateB.order, fillB, orderStateA.order, fillA);
-            fillA.B = fillB.S;
+            fillB.S = fillA.B;
```

Comment:
Is this a typo? I thought we should always use maker's Buy amount?

Reply:
Oh, It turns out that buy order matches with maker's price and sale order matches with taker's price, isn't it?

Reply:
It always matches with the maker order's price, no matter what. When the taker order is a buy order the taker pays less (in tokenS) than the taker order's rate, when it's a sell order the taker buys more (in tokenB) than the taker order's rate. The exchange rate is exactly the rate of the maker's order in both cases.

Reply:
Hi, Brechet, could you please help to check this case
```
    SECTION("Regression Case 2, maker taker role test")
    {
        OrderState makerA = setOrderState(
            orderStateA,
            A_orderID,
            FieldT("2"), FieldT("2000"), false,
            FieldT("10000"),
            0, false, A_orderID
        );
        OrderState takerB = setOrderState(
            orderStateB,
            B_orderID,
            FieldT("4000"), FieldT("1"), true,
            FieldT("10000"),
            0, false, B_orderID
        );
        FieldT expectedFillS_A("1");
        FieldT expectedFillS_B("4000");
        orderMatchingChecked(
            exchangeID, timestamp,
            makerA, takerB,
            true, ExpectedValid::Valid,
            ExpectedFill::Manual, expectedFillS_A, expectedFillS_B
        );
    } 
```
The case is:
Before Order

| Order | Maker A | Taker B |
| ------- | -------- | ------- |
| Sale |      2 iPhones   |   4000 USD  |
| Buy |   2000 USD     |       1 iPhone  |

Maker A's price is 1000USD/iPhone, Taker B's price is 4000USD/iPhone, so these 2 are matchable.  If use maker's price to make the deal. We should get

| Deal | Maker A | Taker B |
| ------- | -------- | ------- |
| Sale |      1 iPhones   |   1000 USD  |
| Buy |   1000 USD     |       1 iPhone  |

But from the case, we can see the real result is

| Deal | Maker A | Taker B |
| ------- | -------- | ------- |
| Sale |      1 iPhones   |   4000 USD  |
| Buy |   4000 USD     |       1 iPhone  |

Is there anything wrong with my understanding? Thanks!

The only thing I thought may be related to this case is you mentioned A is always the Taker before, but from code, the A is the maker if A places a sale order.
```
        if (orderStateA.order.buy == FieldT::one())
        {
            matchable = match(orderStateA.order, fillA, orderStateB.order, fillB);
            fillA.S = fillB.B;
        }
        else
        {
            // here A is maker according to declaration of match().
            matchable = match(orderStateB.order, fillB, orderStateA.order, fillA);
            fillA.B = fillB.S;
        }
```

Reply:
> Is there anything wrong with my understanding? Thanks!

orderA is the taker, orderB is the maker. Always, no exceptions. The orders are switched in `match` but that does NOT make orderA the maker order and orderB the taker order. That's just how the matching code is implemented, it does not change which order is the taker/maker.

The rate of the maker order (orderB!) is 4000/1 so that's the rate that is used in the settlement. orderA, the taker order, is a sell order, so can receive more tokenB (the taker's order is just 1/1000).

> The only thing I thought may be related to this case is you mentioned A is always the Taker before, but from code, the A is the maker if A places a sale order.
> 
> ```
>         if (orderStateA.order.buy == FieldT::one())
>         {
>             matchable = match(orderStateA.order, fillA, orderStateB.order, fillB);
>             fillA.S = fillB.B;
>         }
>         else
>         {
>             // here A is maker according to declaration of match().
>             matchable = match(orderStateB.order, fillB, orderStateA.order, fillA);
>             fillA.B = fillB.S;
>         }
> ```

No, A is still the taker and B is still the maker in BOTH cases. This is just the way I implemented it to reuse as much code as possible.

btw, if we're going to use the `checkFillRate` only method we don't have to do worry about this any more. :) The relayer could do whatever it wants as long as the fill rate is <= the order's rate. The taker/maker model would just be one of the ways to do the matching.

Reply:
Thanks Brechet, I will double check this arduous role-playing game. :)
Just talked with Relay guys, they would like to offer FillS_A/B using a sort of 2-pass method which makes encoded FillS_A/B pass all checkings to avoid such precision issue, I think that's better, let me know if you have more concerns. Thanks!